public/app/plugins/datasource/prometheus/datasource.ts

```diff
@@ -108,10 +108,16 @@ export class PrometheusDatasource {
   }
 
   clampRange(start, end, step) {
-    return {
-      start: start - start % step,
-      end: end - end % step + step,
-    };
+    const startJitter = start % step;
+    const endJitter = end % step;
+    // Shift interval forward on jitter
+    if (startJitter || endJitter) {
```

Comment:
Well, if you want to mirror Prometheus in the number of points, you can always round the range down rather than up (because that's essentially what Prometheus does). I was instead going for covering the whole requested range (focusing particularly on the end of the range so you don't withhold that information until the a whole `step` has passed).

The deeper problem I noticed was that your code will indeed return 2 data points with `step = 3`, `start = 1` and `end = 5`; but one second later -- when `start = 2` and `end = 6` -- it will return 3 data points: `3`, `6` and `9`, with the former leaving out the data at `2`, which was "requested" and the latter unlikely to ever have any data (assuming the wall time is `6`).